torch/distributions/utils.py

```diff
@@ -11,3 +11,55 @@ def expand_n(v, n):
         return torch.Tensor([v]).expand(n, 1)
     else:
         return v.expand(n, *v.size())
+
+
+def _broadcast_shape(shapes):
+    """
+    Given a list of tensor sizes, returns the size of the resulting broadcasted
+    tensor.
+
+    Args:
+        shapes (list of torch.Size): list of tensor sizes
+    """
+    shape = torch.Size([1])
+    for s in shapes:
+        shape = torch._C._infer_size(s, shape)
+    return shape
+
+
+def broadcast_all(*values):
+    """
+    Given a list of values (possibly containing numbers), returns a list where each
+    value is broadcasted based on the following rules:
+      - `torch.Tensor` and `torch.autograd.Variable` instances are broadcasted as
+        per the `broadcasting rules
+        <http://pytorch.org/docs/master/notes/broadcasting.html>`_
+      - numbers.Number instances (scalars) are upcast to Tensor/Variable having
+        the same size and type as the first tensor passed to `values`. If all the
+        values are scalars, then they are upcasted to `torch.Tensor` having size
+        `(1,)`.
+
+    Args:
+        values (list of `numbers.Number`, `torch.autograd.Variable` or
+        `torch.Tensor`)
+
+    Raises:
+        ValueError: if any of the values is not a `numbers.Number`, `torch.Tensor`
+            or `torch.autograd.Variable` instance
+    """
+    scalars = [(idx, v) for idx, v in enumerate(values) if isinstance(v, Number)]
```

Comment:
Wow this turns out to be tricky. I wonder if it's simpler to special-case Variable vs Tensor vs Number?
```py
# Promote numbers to Variables or Tensors.
if any(isinstance(v, Variable) for v in values):
    template = next(v.data for v in values if isinstance(v, Variable))
    values = [Variable(template.new([v])) if isinstance(v, Number) else v for v in values]
elif any(is_tensor(v) for v in values):
    template = next(v for v in values if is_tensor(v))
    values = [template.new([v]) if isinstance(v, Number) else v for v in values]
else:  # Everything is a number.
    values = [Tensor([v]) for v in values]

# Broadcast all shapes. 
shape = _broadcast_shape([v.size() for v in values])
return [v.expand(shape) for v in values]
```
I believe it also saves memory to use `template.new([v])` rather than `template.new(shape).fill_(v)`.

Reply:
Yes, will make the optimization, and let me see if I can simplify this further.

Reply:
Removed the list comprehensions and the zipping logic in favor of two cases, and am just using simple for loops now, since we don't really need to create new lists and it is just easier to operate by tracking scalars / tensors via indices.